tests/kernel/smp/src/main.c

```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
+
+		sync_lock(&key);
+
+		global_cnt++;
+		global_cnt--;
+		global_cnt++;
+
+		sync_unlock(&key);
+	}
+}
+
+static int run_concurrency(int type, void *func)
+{
+	uint32_t start_t, end_t;
+
+	sync_init(type);
+	global_cnt = 0;
+	start_t = k_cycle_get_32();
+
+	tinfo[0].tid =
+	k_thread_create(&tthread[0], tstack[0], STACK_SIZE,
+			(k_thread_entry_t)func,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	tinfo[1].tid =
+	k_thread_create(&tthread[1], tstack[1], STACK_SIZE,
+			(k_thread_entry_t)func,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_tid_t tid =
+	k_thread_create(&t2, t2_stack, T2_STACK_SIZE,
+			(k_thread_entry_t)func,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_thread_join(tinfo[0].tid, K_FOREVER);
+	k_thread_join(tinfo[1].tid, K_FOREVER);
+	k_thread_join(tid, K_FOREVER);
+	cleanup_resources();
+
+	end_t =  k_cycle_get_32();
+
+	printk("type %d: cnt %d, spend %u ms\n", type, global_cnt,
+		k_cyc_to_ms_ceil32(end_t - start_t));
+
+	return global_cnt == (LOOP_COUNT * 3);
+}
+
+/**
+ * @brief Test if the concurrency of SMP works or not
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate the global lock and unlock API of SMP are thread-safe.
+ * We make 3 thread to increase the global count in differenet cpu and
+ * they both do locking then unlocking for LOOP_COUNT times. It shall be no
+ * deadlock happened and total global count shall be 3 * LOOP COUNT.
+ *
+ * We show the 4 kinds of scenairo:
+ * - No any lock used
+ * - Use global irq lock
+ * - Use semaphore
+ * - Use mutex
+ */
+void test_inc_concurrency(void)
+{
+	/* increasing global var without lock */
+	zassert_false(run_concurrency(LOCK_NO, inc_global_cnt),
+			"total count %d is wrong", global_cnt);
+
+	/* increasing global var with irq lock */
+	zassert_true(run_concurrency(LOCK_IRQ, inc_global_cnt),
+			"total count %d is wrong(i)", global_cnt);
+
+	/* increasing global var with irq lock */
+	zassert_true(run_concurrency(LOCK_SEM, inc_global_cnt),
+			"total count %d is wrong(s)", global_cnt);
+
+	/* increasing global var with irq lock */
+	zassert_true(run_concurrency(LOCK_MUTEX, inc_global_cnt),
+			"total count %d is wrong(M)", global_cnt);
+}
+
+struct _cpu *running_cpu;
+
+static void tIsr_current_cpu(const bool *has_locked)
+{
+	int k = 0;
+
+	if (*has_locked) {
+		k = arch_irq_lock(); }
+
+	running_cpu = _current_cpu;
+
+	if (*has_locked) {
+		arch_irq_unlock(k); }
+}
+
+/**
+ * @brief Test inadequate usage of _current_cpu causing assertion
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details The _current_cpu macro can be only used in neigther
+ * interrupt context nor interrupt locked status, otherwise an
+ * assertion will be triggered. Because current context can be
+ * preempted and migrated to another SMP CPU during getting the
+ * current CPU. Using this is more safe than arch_curr_cpu() in
+ * context switch code.
```

Comment:
We're writing a test to validate an assertion?  Isn't that circular?  Don't we now need a separate test case to validate the test that validates the assertion?  Only then can we get started on the test to validate the test case to validate the test that validates the assertion.

Just remove.  That assertion is there and works to protect against misuse.  And I know because I misuse the API all the time and trip over it.

Reply:
Got it, @andyross I removed this test case. But here I still try to cover the branch of _current_cpu. So I just try to put the validation of the assertion in another existing testcase test_get_cpu(), do you think is this acceptable?  Thank you!

Reply:
Again, this is being IMHO senselessly strict about code coverage.  An assertion is something that should "never happen".  So **by definition**, it represents uncovered code, and it always will.  All assertions work like that.  We'll spend forever chasing them down trying to "cover" them, and that's completely missing the point.

Reply:
Thanks for detailing this concept, so helpful. I think this corrects some of my misunderstandings. I just removed them completely.